### DISCUSSION TITLE
Questions & musings

### DISCUSSION THREAD

**Discussion by immuneeb** [lzrscg/concorde#3] [Q&A] [answered]
Just read the project [overview doc](https://docs.google.com/document/d/1AXojYg9ZRrPU9UzUTVMs_WWszt7dD9aZkxGTY3alZd4/edit#heading=h.21fytwovsntk) - very clear & thoughtful! 

main question - why is it required for everyone in a group to agree on the same app version?

think this requirement raises two tradeoffs -

1. **Value of forking** - switching costs increase exponentially if i have to convince others to switch to a fork i like vs. just doing so myself. this means forks may have a harder time capturing traction, which in turn means that "major" branches could grab smaller forks' code without compensating them. ultimately this could reduce the incentive to fork / permissionless innovate.
2. **UX impact** - would be annoying if i'm in 10 different groups & have to manage 10 different versions of the app.

curious if there are other variations we could consider with different tradeoffs. in particular think ability of forks to fairly capture value is really important to the vision of rapidly evolving app!

**Comment by lzrscg** [comment, marked as answer]
First of all, thanks for reading the doc! This is a really great question.

### Short Answer
> Concorde is making a conscious trade off that instead of developers having to coordinate app compatibility between all versions (a very hard problem at scale), small groups just need to decide which version of the app to use internally. This is a good system since central coordination works well for small groups.
>
> To be clear, I never meant to suggest that everyone would have to independently decide to download the same version of the app. You don't need to convince your friends, per se. The group admins select the version.
>
> In the future, Concorde will allow groups to run multiple versions simultaneously. People within a group can vote with their feet and use each one as much or as little as they want.

### Explanation
Everyone in the group must agree on the same version of the app because...
1. Using different clients can create [fragmentation](https://twitter.com/matrixdotorg/status/1477983547856232452). Fragmentation is where two different versions of the app try to handle messages differently, which can cause errors for users or other unexpected outcomes. If everyone in the group uses the same version, this is not an issue.
2. Central coordination is optimal for small groups, but slows development at scale (see image below). Concorde is making a conscious trade off that **instead of developers having to coordinate app compatibility between all versions** (a very hard problem at scale), **small groups just need to decide which version of the app to use**.
<img width="690" alt="Screen Shot 2022-01-04 at 9 40 36 AM" src="https://user-images.githubusercontent.com/84213303/148988101-51911857-f237-4d2a-9ab4-e6b77f3a5cde.png">

If you think about it, most groups are typically fine with having an admin select the app version. For example, I am the admin of this GitHub repo and I enabled the "Discussions" feature we are using right now. This actually is a good system since, as I mentioned above, central coordination works well for small groups. **To be clear, I never meant to suggest that everyone would have to independently decide to download the same version of the app.** You don't need to convince your friends, per se. The group admin selects the version.

Ultimately, Concorde is really a general app platform. In the future, Concorde will allow groups to run multiple versions simultaneously. There could be a version for **chat** and a version that is a **game**. Users could switch between these experiences using an _in-group app launcher_. This would mean that within a group, no one has to decide who runs which app. People can vote with their feet and use each one as much or as little as they want. But the important part is that different versions of the app never communicate with each other because this causes fragmentation, and—in turn—creates coordination overhead for developers which hurts the scalability of the iteration process.
